Add support to coerce structural type

Yaliang opened the pull request:
Add support to coerce structural type:

1. Nested primitive/structural field coercion
2. Adding/removing tailing fields in (nested) struct field

+cc @dabaitu

Yaliang commented:
@geraint0923 I noticed you initially implement the coercion for primitive types. Could you advise who can take a review for this patch?

DaimonPl commented:
I really appreciate this PR.

We are strugling with schema evolution problems in our infrastructure. This PR looks to be solving them all :)

I hope it won't be skipped like some older PRs :)

martint commented:
@haozhun, can you review this?

losipiuk commented:
@martint, @haozhun just reminding myself about ancient PR which does a bit of refactor and simplification of coercion mechanism in hive connector. As well as adding support for decimal types. 

I've just rebased it to current master and new incarnation is here: https://github.com/prestodb/presto/pull/9409.

It would be great if this one is put on top of refactorings from my PR (assuming we want those).

martint commented:
> I've just rebased it to current master and new incarnation is here: #9409.

Did you mean https://github.com/prestodb/presto/pull/9422 ?

losipiuk commented:
> Did you mean #9422

Yeah - corrected

Yaliang commented:
@losipiuk I think there won't be much difficulty to refactor this PR later against your refactorings. There is no interface/public method/class change in this PR. All added methods/classes are private in the existing class, so there won't have hard conflicts for further refactoring. I would suggest we can work on this first and then work on the refactoring as a whole if that's necessary. Thoughts ?

losipiuk commented:
> @losipiuk I think there won't be much difficulty to refactor this PR later against your refactorings.

I am a bit concerned mostly about changes in test as I made a bunch of changes there so much less code needs to be written. I do not expect much work in resolving conflicts of coercion code itself.

Yaliang commented:
I'm happy to drive it through. I will address comment before the end of weed

Yaliang commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercionPolicy.java:
```diff
@@ -62,6 +70,49 @@ public boolean canCoerce(HiveType fromHiveType, HiveType toHiveType)
+        List<HiveType> fromFieldTypes = getAllStructFieldTypeInfos(fromHiveType);
+        List<HiveType> toFieldTypes = getAllStructFieldTypeInfos(toHiveType);
+        for (int i = 0; i < Math.min(fromFieldTypes.size(), toFieldTypes.size()); i++) {
+            if (!fromFieldTypes.get(i).equals(toFieldTypes.get(i)) && !canCoerce(fromFieldTypes.get(i), toFieldTypes.get(i))) {
```
Addressed as the option 1

Yaliang commented:
@haozhun I think it's ready for another round review.
The name check added in the `HiveCoercionPolicy`. 
Used `ColumnarArray`, `ColumnarMap`, and `ColumnarRow` in page-based coercers. 
`mapType` `arrayType` and `rowType` in `HiveTestUtils` are used to format the testing schema.

Thanks for the great advice!

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercionPolicy.java:
```diff
@@ -62,6 +69,45 @@ public boolean canCoerce(HiveType fromHiveType, HiveType toHiveType)
+        List<String> toFieldNames = getAllStructFieldNames(toHiveType);
+        List<HiveType> fromFieldTypes = getAllStructFieldTypeInfos(fromHiveType);
+        List<HiveType> toFieldTypes = getAllStructFieldTypeInfos(toHiveType);
+        for (int i = 0; i < Math.min(fromFieldTypes.size(), toFieldTypes.size()); i++) {
```
In general, we statically import `Math.min`.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:
```diff
@@ -830,4 +832,16 @@ public static void closeWithSuppression(RecordCursor recordCursor, Throwable thr
         }
     }
+
+    public static List<HiveType> getAllStructFieldTypeInfos(HiveType hiveType)
```
`getAllStructFieldTypeInfos` is a particularly bad name because it's the same as the one you are using 2 lines below, but it's something quite different.

I recommend inlining this function and the one below.

Yaliang replied:
It's referred to multiple places like the coercers for structs. I agree that the name is confusing. I would rename it as `getAllStructFieldHiveTypes`.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:
```diff
@@ -830,4 +832,16 @@ public static void closeWithSuppression(RecordCursor recordCursor, Throwable thr
+    {
+        return ((StructTypeInfo) hiveType.getTypeInfo()).getAllStructFieldTypeInfos().stream()
+                .map(typeInfo -> HiveType.valueOf(typeInfo.getTypeName()))
+                .collect(Collectors.toList());
```
In general, we use `ImmutableList.toImmutableList` in Guava instead. Also, we generally statically import this function whenever we use it.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java:
```diff
@@ -20,17 +20,23 @@
 import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.facebook.presto.spi.block.ColumnarArray;
+import com.facebook.presto.spi.block.ColumnarMap;
+import com.facebook.presto.spi.block.ColumnarRow;
```
Statically import the utility methods you used from these three classes, and remove import of the classes themselves.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java:
```diff
@@ -287,6 +294,15 @@ else if (fromHiveType.equals(HIVE_INT) && toHiveType.equals(HIVE_LONG)) {
         else if (fromHiveType.equals(HIVE_FLOAT) && toHiveType.equals(HIVE_DOUBLE)) {
             return new FloatToDoubleCoercer();
         }
+        else if (HiveUtil.isArrayType(fromType) && HiveUtil.isArrayType(toType)) {
```
statically import these `HiveUtil` methods

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarArray.java:
```diff
@@ -34,7 +37,7 @@ public static ColumnarArray toColumnarArray(Block block)
 
         if (!(block instanceof AbstractArrayBlock)) {
-            throw new IllegalArgumentException("Invalid array block");
+            throw new IllegalArgumentException("Invalid array block: " + block.getClass().getName());
```
Do the same in `ColumnarRow`

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarArray.java:
```diff
@@ -26,6 +26,9 @@ public static ColumnarArray toColumnarArray(Block block)
     {
         requireNonNull(block, "block is null");
 
+        if (block instanceof LazyBlock) {
```
Remove handling of `LazyBlock` from all three.

Instead, make `CoercionLazyBlockLoader` strip the `LazyBlock` layer when necessary.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java:
```diff
@@ -424,6 +440,163 @@ public Block apply(Block block)
+            if (elementCoercer == null) {
+                return block;
+            }
+            BlockBuilder blockBuilder = toType.createBlockBuilder(new BlockBuilderStatus(), block.getPositionCount());
```
In all 3, do not use `BlockBuilder`.

Instead, new `new ArrayBlock`, `mapType.createBlockFromKeyValue`, and `new RowBlock`. This is the same way they are done in XxxStreamReader in presto-orc.

This is likely much easier to implement as well.

Performance is critical HivePageSource as ORC, Parquet, DWRF goes through this path. In a typical Presto set up, more than 80% of time is spent in TableScanOperator and ScanFilterProjectOperator.

Yaliang replied:
I tried to refactor `ListCoercer`. But I realize it's not visible to get the offsets of coerced element block. Any idea?

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:
```diff
@@ -830,4 +832,11 @@ public static void closeWithSuppression(RecordCursor recordCursor, Throwable thr
         }
     }
+
+    public static List<HiveType> getAllStructFieldHiveTypes(HiveType hiveType)
```
Let's call this `extractStructFieldTypes`

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercionPolicy.java:
```diff
@@ -62,6 +70,45 @@ public boolean canCoerce(HiveType fromHiveType, HiveType toHiveType)
+        HiveType fromValueType = HiveType.valueOf(((MapTypeInfo) fromHiveType.getTypeInfo()).getMapValueTypeInfo().getTypeName());
+        HiveType toKeyType = HiveType.valueOf(((MapTypeInfo) toHiveType.getTypeInfo()).getMapKeyTypeInfo().getTypeName());
+        HiveType toValueType = HiveType.valueOf(((MapTypeInfo) toHiveType.getTypeInfo()).getMapValueTypeInfo().getTypeName());
+        return (fromKeyType.equals(toKeyType) || canCoerce(fromKeyType, toKeyType)) && (fromValueType.equals(toValueType) || canCoerce(fromValueType, toValueType));
```
This did not exceed Presto line width limit. However, for readability, reformat as:

```
        return (fromKeyType.equals(toKeyType) || canCoerce(fromKeyType, toKeyType)) &&
                (fromValueType.equals(toValueType) || canCoerce(fromValueType, toValueType));

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercionPolicy.java:
```diff
@@ -62,6 +70,45 @@ public boolean canCoerce(HiveType fromHiveType, HiveType toHiveType)
+        List<HiveType> fromFieldTypes = getAllStructFieldHiveTypes(fromHiveType);
+        List<HiveType> toFieldTypes = getAllStructFieldHiveTypes(toHiveType);
+        for (int i = 0; i < min(fromFieldTypes.size(), toFieldTypes.size()); i++) {
+            if (!fromFieldNames.get(i).equals(toFieldNames.get(i)) || (!fromFieldTypes.get(i).equals(toFieldTypes.get(i)) && !canCoerce(fromFieldTypes.get(i), toFieldTypes.get(i)))) {
```
For readability, do

```
            if (!fromFieldNames.get(i).equals(toFieldNames.get(i))) {
                return false;
            }
            if (!fromFieldTypes.get(i).equals(toFieldTypes.get(i)) && !canCoerce(fromFieldTypes.get(i), toFieldTypes.get(i))) {
                return false;
            }

Yaliang commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercionRecordCursor.java:
```diff
@@ -367,4 +386,297 @@ public void coerce(RecordCursor delegate, int field)
+            int field)
+    {
+        Object value = null;
+        if (fromFieldHiveType.equals(HIVE_BYTE) || fromFieldHiveType.equals(HIVE_SHORT) || fromFieldHiveType.equals(HIVE_INT) || fromFieldHiveType.equals(HIVE_LONG) || fromFieldHiveType.equals(HIVE_FLOAT)) {
```
How could we test boolean and double type? We wouldn't even reach this method in the current schema mismatched table tests.

Yaliang commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercionRecordCursor.java:
```diff
@@ -367,4 +386,297 @@ public void coerce(RecordCursor delegate, int field)
+            Coercer coercer,
+            int field)
+    {
+        Object value = null;
```
I eventually used a single `BridgingRecordCursor` per `HiveCoercionRecordCursor` and de-referring the `value` variable at the end of `rewriteBlock`. I don't know if we actually need to set the correct `Type` here since we wouldn't visit `getType` method. What's your opinion?

Yaliang commented:
@haozhun I addressed most of the comments except the one about testing all 5 Presto types and ready for another round of review(The travis CI is passed in my repo's branch, FYI). About the ask about the testing, do we want to test it specifically on the `rewriteBlock` method or just add some exception tests on unsupported coercion(which I think may not actually reach `rewriteBlock` method)?

haozhun commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveCoercion.java:
```diff
@@ -94,31 +108,16 @@
+                        format("    float_to_double            %s,", floatToDoubleType) +
+                        format("    row_to_row                 STRUCT < tinyint_to_smallint : TINYINT, tinyint_to_int : TINYINT, tinyint_to_bigint : TINYINT, smallint_to_int : SMALLINT, smallint_to_bigint : SMALLINT, int_to_bigint : INT, bigint_to_varchar : BIGINT, float_to_double : %s >,", floatToDoubleType) +
+                        format("    list_to_list               ARRAY < STRUCT < tinyint_to_smallint : TINYINT, tinyint_to_int : TINYINT, tinyint_to_bigint : TINYINT, smallint_to_int : SMALLINT, smallint_to_bigint : SMALLINT, int_to_bigint : INT, bigint_to_varchar : BIGINT, float_to_double : %s > >,", floatToDoubleType) +
+                        format("    map_to_map                 MAP < TINYINT, STRUCT < tinyint_to_smallint : TINYINT, tinyint_to_int : TINYINT, tinyint_to_bigint : TINYINT, smallint_to_int : SMALLINT, smallint_to_bigint : SMALLINT, int_to_bigint : INT, bigint_to_varchar : BIGINT, float_to_double : %s > >", floatToDoubleType) +
```
Your test is very complete. However, I find it a bit too hard to read.

For readability, I make a few proposals:
* Use abbreviations. (We avoid abbreviation in Presto in general. But I find it excessively wordy in this case otherwise)
* Avoid unnecessary spaces.
* Don't cover all conversions in each of row_to_row, list_to_list, map_to_map. Instead, cover all of them across the three. (And add comment to point this out.)

In addition, I find it useful to have non coerced items in the structs. In your implementation, they go to quite different code paths.

```
// all primitive/varchar coercions are covered across row_to_row, list_to_list, and map_to_map 
"    row_to_row                 STRUCT<ti2si: TINYINT, si2int: SMALLINT, int2bi: INT, bi2vc: BIGINT>," +
"    list_to_list               ARRAY<STRUCT<ti2int: TINYINT, si2bi: SMALLINT, bi2vc: BIGINT>>," +
"    map_to_map                 MAP<TINYINT, STRUCT<ti2bi: TINYINT, int2bi: INT, float2double: " + floatToDoubleType + ">>" +

haozhun commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveCoercion.java:
```diff
@@ -259,20 +269,37 @@ public void testHiveCoercionAvro()
                         "(-1, 2, -3, 100, -101, 2323, 12345, 0.5)," +
                         "(1, -2, null, -100, 101, -2323, -12345, -1.5)",
-                tableName));
+                dummyTableName));
```
I find the cost you pay in order to avoid duplication in the `VALUES` clause too high. By introducing the dummy table, a nontrivial amount of complexity is introduced to this test. It makes the test harder to read and understand.

Instead, I would manually write it out, and remove the dummy table:
```
                        "(-1, 2, -3, 100, -101, 2323, 12345, 0.5," +
                        "    named_struct(...)," +
                        "    array(named_struct(...))," +
                        "    map(..., named_struct(...)))," +
                        "(1, -2, null, -100, 101, -2323, -12345, -1.5," +
                        "    named_struct(...)," +
                        "    array(named_struct(...))," +
                        "    map(..., named_struct(...)))," +
```

Yaliang replied:
Just realized I have tried this before and it seems impossible.

    Unable to create temp file for insert values Expression of type TOK_FUNCTION not supported in insert/values

haozhun replied:
There is actually no reason that these data is inserted with Hive. (Maybe there was at the time.) I changed it to use Presto, which offers the necessary SQL expressiveness.

haozhun commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveCoercion.java:
```diff
@@ -351,6 +396,9 @@ private static void alterTableColumnTypes(String tableName)
         executeHiveQuery(format("ALTER TABLE %s CHANGE COLUMN float_to_double float_to_double double", tableName));
+        executeHiveQuery(format("ALTER TABLE %s CHANGE COLUMN row_to_row row_to_row struct<tinyint_to_smallint:smallint,tinyint_to_int:int,tinyint_to_bigint:bigint,smallint_to_int:int,smallint_to_bigint:bigint,int_to_bigint:bigint,bigint_to_varchar:string,float_to_double:double>", tableName));
+        executeHiveQuery(format("ALTER TABLE %s CHANGE COLUMN list_to_list list_to_list array<struct<tinyint_to_smallint:smallint,tinyint_to_int:int,tinyint_to_bigint:bigint,smallint_to_int:int,smallint_to_bigint:bigint,int_to_bigint:bigint,bigint_to_varchar:string,float_to_double:double>>", tableName));
+        executeHiveQuery(format("ALTER TABLE %s CHANGE COLUMN map_to_map map_to_map map<int,struct<tinyint_to_smallint:smallint,tinyint_to_int:int,tinyint_to_bigint:bigint,smallint_to_int:int,smallint_to_bigint:bigint,int_to_bigint:bigint,bigint_to_varchar:string,float_to_double:double>>", tableName));
```
ditto here. (See my comment in `tableDefinitionBuilder` in AbstractTestHiveClient)

haozhun commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveCoercion.java:
```diff
@@ -293,6 +320,15 @@ private void doTestHiveCoercion(HiveTableDefinition tableDefinition)
+                queryResult.column(10).stream().map(o -> Arrays.asList((Object[]) ((PrestoArray) o).getArray())).collect(Collectors.toList()),
+                rowColumn.stream().map(ImmutableList::of).collect(Collectors.toList()),
+                "list_to_list field is not equal");
+        assertEqualsIgnoreOrder(queryResult.column(11), rowColumn.stream().map(map -> ImmutableMap.of(map.get("tinyint_to_int"), map)).collect(Collectors.toList()), "map_to_map field is not equal");
```
Remove this block of code. Instead, write them out directly in the lines above (without any logic).

Yaliang replied:
Realized `assertThat` don't support `JAVA_OBJECT` directly. Has to extract each nested field. 

```
java.lang.RuntimeException: Unsupported sql type JAVA_OBJECT
	at io.prestodb.tempto.internal.query.QueryResultValueComparator.compare(QueryResultValueComparator.java:108)
	at io.prestodb.tempto.assertions.QueryAssert.isAnyValueEqual(QueryAssert.java:344)
	at io.prestodb.tempto.assertions.QueryAssert.rowsEqual(QueryAssert.java:334)
	at io.prestodb.tempto.assertions.QueryAssert.containsRow(QueryAssert.java:316)
	at io.prestodb.tempto.assertions.QueryAssert.contains(QueryAssert.java:176)
	at io.prestodb.tempto.assertions.QueryAssert.containsOnly(QueryAssert.java:207)
	at io.prestodb.tempto.assertions.QueryAssert.containsOnly(QueryAssert.java:219)
	at com.facebook.presto.tests.hive.TestHiveCoercion.doTestHiveCoercion(TestHiveCoercion.java:285)
	at com.facebook.presto.tests.hive.TestHiveCoercion.testHiveCoercionOrc(TestHiveCoercion.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

haozhun commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveCoercion.java:
```diff
@@ -386,4 +434,14 @@ private static QueryResult executeHiveQuery(String query)
         return testContext().getDependency(QueryExecutor.class, "hive").executeQuery(query);
     }
+
+    private static Map<Object, Object> namedStruct(Object... objects)
```
With `ImmutableMap.of(...)`, you don't need this helper (up to 5 key-value entries).

When you go over, we can't do this any more. But we wouldn't do it anyways because readability is pretty bad at that point anyways. Instead, we do:
```
ImmutableMap.<K, V>builder()
        .put(..., ...)
        .put(..., ...)
        .put(..., ...)
        .put(..., ...)
        .put(..., ...)
        .build()

Yaliang replied:
I initially come across `ImmutableMap` but then I realized I couldn't use it because I want to test `null` value but `ImmutableMap` don't allow `null`.

haozhun commented:
Please address my comments, and rebase on top of latest master. Thank you, @Yaliang.

The code is in generally good shape. And I think we will be able to merge to this in the next week or the week after.

haozhun commented:
I made edits to a comment. Please use the latest version on github, and avoid reading the email notification from github.

Yaliang commented:
I have addressed most comments, rebased on top of master, tried best to make it easy to read and ready for another round of review (I wish it's the final one). Thanks for all the feedbacks through, @haozhun.

It seems to be impossible to avoid a dummy table. But I avoid putting meaningful values so it would be easy to understand now. `QueryAssert` in `tempto` can't compare `JAVA_OBJECT`, so I write some helper functions to extract/project the value from `QueryResult` and the list of expected `Row`s.

Yaliang commented:
Rebased again

haozhun commented:
I made quite a few cleanups in `TestHiveCoercion`. I added comments to HiveCoercionPolicy to explain what we allow.

I will merge this with my changes after 0.195 is out. Schema evolution for row types in Hive connector will be available in 0.196.

haozhun commented:
Merged #9131
